dev/benchmarks/macrobenchmarks/test/util.dart

```diff
@@ -0,0 +1,192 @@
+    }
+    final List<int> frameBuildTimeMicros = data.map<int>(
+      (FrameTiming datum) => datum.buildDuration.inMicroseconds,
+    ).toList();
+    final double averageFrameBuildTime =
+        frameBuildTimeMicros.reduce((int a, int b) => a+b) / 1E3 / data.length;
+    final List<int> frameBuildTimeMicrosSorted =
+        List<int>.from(frameBuildTimeMicros)..sort();
+    final double percentileFrameBuildTime90 =
+        _findPercentile(frameBuildTimeMicrosSorted, 0.90) / 1E3;
+    final double percentileFrameBuildTime99 =
+        _findPercentile(frameBuildTimeMicrosSorted, 0.99) / 1E3;
+    final double worstFrameBuildTime = frameBuildTimeMicrosSorted.last / 1E3;
+    final int missedFrameBuildBudget =
+        _countExceed(frameBuildTimeMicrosSorted, kBuildBudget.inMicroseconds);
+
+    final List<int> frameRasterizerTimeMicros = data.map<int>(
+      (FrameTiming datum) => datum.rasterDuration.inMicroseconds,
+    ).toList();
+    final double averageFrameRasterizerTime =
+        frameRasterizerTimeMicros.reduce((int a, int b) => a+b) / 1E3 / data.length;
+    final List<int> frameRasterizerTimeMicrosSorted =
+        List<int>.from(frameBuildTimeMicros)..sort();
+    final double percentileFrameRasterizerTime90 =
+        _findPercentile(frameRasterizerTimeMicrosSorted, 0.90) / 1E3;
+    final double percentileFrameRasterizerTime99 =
+        _findPercentile(frameRasterizerTimeMicrosSorted, 0.90) / 1E3;
+    final double worstFrameRasterizerTime = frameRasterizerTimeMicrosSorted.last / 1E3;
+    final double missedFrameRasterizerTime =
+        _countExceed(frameRasterizerTimeMicrosSorted, kBuildBudget.inMicroseconds) / 1E3;
+
+    return <String, dynamic>{
+      'average_frame_build_time_millis': averageFrameBuildTime,
+      '90th_percentile_frame_build_time_millis': percentileFrameBuildTime90,
+      '99th_percentile_frame_build_time_millis': percentileFrameBuildTime99,
+      'worst_frame_build_time_millis': worstFrameBuildTime,
+      'missed_frame_build_budget_count': missedFrameBuildBudget,
+      'average_frame_rasterizer_time_millis': averageFrameRasterizerTime,
+      '90th_percentile_frame_rasterizer_time_millis': percentileFrameRasterizerTime90,
+      '99th_percentile_frame_rasterizer_time_millis': percentileFrameRasterizerTime99,
+      'worst_frame_rasterizer_time_millis': worstFrameRasterizerTime,
+      'missed_frame_rasterizer_budget_count': missedFrameRasterizerTime,
+      'frame_count': data.length,
+      'frame_build_times': frameBuildTimeMicros,
+      'frame_rasterizer_times': frameRasterizerTimeMicros,
+    };
+  }
+}
+
+// The following helper functions require data sorted
+
+// return the 100*p-th percentile of the data
+T _findPercentile<T extends num>(List<T> data, double p) {
+  assert(p >= 0 && p <= 1);
+  return data[((data.length - 1) * p).round()];
+}
+
+// return the number of items in data that > threshold
+int _countExceed<T extends num>(List<T> data, T threshold) {
```

Comment:
Nit: I feel it's safer to use a simple O(n) scanning algorithm here than a cleverer O(log n) binary search. We don't have unit test coverage for this function so it might be easy for someone to break this in the future without noticing it.

Reply:
Done. A little sad when I finally find small algos are useful somewhere but it turns out not a good practice :(

Reply:
I feel sorry too... Here's how I think you might be able to add it back in the future:

Writing that algorithm is a good, what feels unsafe is to make it as an private little function without any unit tests. In the future, if `FrameTimingSummarizer` turns out to be something that we want to add to our API, then we can invest a lot more time in it to improve its ergonomics and efficiency. At that time, we shall certainly make the algorithm as efficient as possible, and add many unit tests to guarantee its correctness. (Right now we're just trying to unblock our critical path asap.)